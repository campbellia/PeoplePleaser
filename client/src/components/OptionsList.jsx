import React from 'react';
import { Paper, Table, TableBody, TableRow, TableHead, TableCell, IconButton, Grid, Typography } from '@material-ui/core';
import { Delete } from '@material-ui/icons';
import { makeStyles } from '@material-ui/styles';
import AddOptionsForm from './AddOptionsForm.jsx';

const useStyles = makeStyles(theme => ({
  narrowCell: {
    paddingTop: 0,
    paddingBottom: 0
  }
}));

const OptionsList = ({ options, handleRemoveOption, handleAddOption }) => {
  const classes = useStyles();

  const makeOptionsList = () => {
    if (options.length) {
      return options.map((option, i) => {
        return (
          <TableRow key={i}>
            <TableCell className={classes.narrowCell}>
              <Grid container direction="row" justify="space-between" alignItems="center" aria-label="option">
                <Grid item>
                  {option}
                </Grid>
                <Grid item >
                  <IconButton id={i} onClick={(e) => {handleRemoveOption(e.target.id)}} aria-label="delete">
                    <Delete />
                  </IconButton>
                </Grid>
              </Grid>
            </TableCell>
          </TableRow>);
      });
    } else {
      return (
        <TableRow>
          <TableCell>Add Some Options!</TableCell>
        </TableRow>);
    }
  }

  return (
    <Paper >
      <Table >
        <TableHead>
          <TableRow>
            <TableCell>
              <Typography color="inherit" variant="h6">OPTIONS</Typography>
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {makeOptionsList()}
          <TableRow>
            <TableCell className={classes.narrowCell}>
              <AddOptionsForm handleAddOption={handleAddOption}/>
            </TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </Paper>
  );
}

export default OptionsList;
