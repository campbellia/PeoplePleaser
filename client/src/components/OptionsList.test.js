import React from 'react';
import {render, fireEvent} from '@testing-library/react';
import OptionsList from './OptionsList';

describe('OptionsList', () => {
  const props = {
    options: ["Bananas", "Oranges", "Walnuts", "Grapefruit"],
    handleAddOption: newOption => newOption,
    handleRemoveOption: optionIndex => optionIndex
  }

  const addOptionSpy = jest.spyOn(props, 'handleAddOption');
  const removeOptionSpy = jest.spyOn(props, 'handleRemoveOption');

  it('should display the options passed down from its parent component', () => {
    const { getByText } = render(
      <OptionsList options={props.options} handleAddOption={props.handleAddOption} handleRemoveOption={props.handleRemoveOption} />);
    expect(getByText("Walnuts")).toBeInTheDocument();
    expect(getByText("Oranges")).toBeInTheDocument();
  });

  it('should not let the user add empty options', () => {
    const { getByLabelText } = render(
      <OptionsList options={props.options} handleAddOption={props.handleAddOption} handleRemoveOption={props.handleRemoveOption} />);

    fireEvent.click(getByLabelText('submitOption'));
    expect(addOptionSpy).not.toHaveBeenCalled();
  });

  it('should pass option input to the parent event handler', () => {
    const { getByTestId, getByLabelText } = render(
      <OptionsList options={props.options} handleAddOption={props.handleAddOption} handleRemoveOption={props.handleRemoveOption} />);

    fireEvent.change(getByTestId('option-input'), {target: {value: 'Beets'}})
    fireEvent.click(getByLabelText('submitOption'));
    expect(addOptionSpy).toHaveBeenCalledTimes(1);
    expect(addOptionSpy).toHaveBeenCalledWith('Beets');
  });

  it('should pass the index of a deleted option up to the parent component event handler', () => {
      const { getByLabelText, getAllByLabelText } = render(
      <OptionsList options={props.options} handleAddOption={props.handleAddOption} handleRemoveOption={props.handleRemoveOption} />);

      //each option should have a delete button rendered
      const deleteButtons = getAllByLabelText('delete');
      expect(deleteButtons.length).toEqual(props.options.length);

      fireEvent.click(deleteButtons[2]);
      expect(removeOptionSpy).toHaveBeenCalledTimes(1);
      expect(removeOptionSpy).toHaveBeenCalledWith("2");
    });

});
