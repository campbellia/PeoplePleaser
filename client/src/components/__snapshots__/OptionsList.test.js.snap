// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsList should render correctly 1`] = `
<WithStyles(ForwardRef(Paper))>
  <WithStyles(ForwardRef(Table))>
    <WithStyles(ForwardRef(TableHead))>
      <WithStyles(ForwardRef(TableRow))>
        <WithStyles(ForwardRef(TableCell))>
          <WithStyles(ForwardRef(Typography))
            color="inherit"
            variant="h6"
          >
            OPTIONS
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableHead))>
    <WithStyles(ForwardRef(TableBody))>
      <WithStyles(ForwardRef(TableRow))>
        <WithStyles(ForwardRef(TableCell))>
          Add Some Options!
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableRow))>
        <WithStyles(ForwardRef(TableCell))
          className="makeStyles-narrowCell-1"
        >
          <AddOptionsForm
            handleAddOption={[Function]}
          />
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableBody))>
  </WithStyles(ForwardRef(Table))>
</WithStyles(ForwardRef(Paper))>
`;
